Generate package path as proto go_package option

fischor opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Changes the go-to-protobuf code-generator to write the full Go package
name (path) as go_package file option instead of the base package name.
Regenerates the corresponding generated.proto files.

The `go_package` option should be the full path of the Go package, not just the package (base) name.

See the changed `kubernetes/staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/generator.go ` file, the rest is just the result of code regeneration.

#### Which issue(s) this PR fixes:

Fixes #102808

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

NONE

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

fischor committed:
Generate package path as proto go_package option

Changes the go-to-protobuf code-generator to write the full Go package
name (path) as go_package file option instead of the base package name.
Regenerates the corresponding generated.proto files.

liggitt commented:
/ok-to-test

liggitt commented:
there are a couple non-generated proto files as well at:

* staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
* staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto

should those be fixed as well?

fischor committed:
Use package import path as go_package option

Sets the full Go import path as go_package options instead of the (base)
package name as this is required for external users. Adapts the output
base in the update-generated-runtime-dockerized.sh script to this
change. Regenerates the corresponding proto files.

fischor committed:
Avoid duplicate boilerplate for .pb.go files

fischor commented:
I have updated the `staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto` and `staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto` protos. That required to make a change in the `hack/update-generated-runtime-dockerized.sh ` shell script as well as the output location is the combination of the `--gogo_out` flag and the `go_package` file options.

There is an additional commit https://github.com/kubernetes/kubernetes/pull/105027/commits/39fa3732bb448a5d802f64d2e37c45073747d49b that was required because otherwise additional boilerplate was added with every run of `./hack/update-generated-protobuf.sh`. E.g. if you run it 5 times then you would end up with 5+1 times the boilerplate in the resulting `generated.pb.go` file. That was not happening before I changed the `go_package` options to `k8s.io/...`. This is likely another bug, I am going to open a separate issue for that. However, with that commit in place additional runs of `./hack/update-generated-protobuf.sh` are not introducing several boilerplate snippets into the `.pb.go` files.

liggitt commented on staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/cmd.go:
```diff
@@ -260,7 +260,6 @@ func Run(g *Generator) {
 	if len(g.Conditional) > 0 {
 		fmt.Fprintf(buf, "// +build %s\n\n", g.Conditional)
 	}
-	buf.Write(boilerplate)
```
is this related to the package path change, or did this issue exist at HEAD as well? if this issue already exists at HEAD, let's resolve it in a separate PR

fischor replied:
It doesn't exist at HEAD. If you apply this to HEAD, the boilerplate will be missing when running code generation. So its likely that the changes I have made somewhat lead to this strange behaviour of applying the boilerplate multiple times. 

I need to understand the go-to-protobuf generator a little bit more to fully understand this. Especially where it gets the files from that are "rewritten" is the code above (i.e. if these files already contain the boilerplate or not).

The strange thing is that I have just changed the single line that prints the package path in the go-to-protobuf generator. I will investigate the further within the next days. Lets wait with merging.

howardjohn replied:
Happened to me too in https://github.com/kubernetes/kubernetes/pull/106630#issuecomment-1005817409, no clue why...

caesarxuchao commented:
/assign @liggitt 
Thanks!

/triage accepted

fischor committed:
Adapt to changed protoc-gen-gogo output paths

Since the autogenerated proto files now have their Go import paths as
go_package option set, protoc-gen-gogo is not generating source
relative outputs anymore. The output location depends soley on the Go
import path. This commit ensures the correct file is picked up for
rewriting and adding boilerplate.

fischor commented:
Previously only the package names where set in the `go_package` options.
That way, `protoc-gen-gogo` generates the output files "source relative" (in the same path they originate from, but under the path specified with `--gogo_out`. Here is an example:

```sh
~/Github/kubernetes/kubernetes ❯ awk "NR==29" vendor/k8s.io/api/flowcontrol/v1beta1/generated.proto
option go_package = "v1beta1";
~/Github/kubernetes/kubernetes ❯ protoc -I . -I vendor -I third_party/protobuf --gogo_out=tmpout vendor/k8s.io/api/flowcontrol/v1beta1/generated.proto

vendor/k8s.io/api/flowcontrol/v1beta1/generated.proto:26:1: warning: Import k8s.io/apimachinery/pkg/runtime/schema/generated.proto is unused.
vendor/k8s.io/api/flowcontrol/v1beta1/generated.proto:25:1: warning: Import k8s.io/apimachinery/pkg/runtime/generated.proto is unused.
~/Github/kubernetes/kubernetes ❯ tree tmpout
tmpout
└── vendor
    └── k8s.io
        └── api
            └── flowcontrol
                └── v1beta1
                    └── generated.pb.go

5 directories, 1 file
```

Now the generated `.proto` files have the full Go import path specified in `go_package` options. `protoc-gen-go` will generate the files then in the folder specified by `go_package` path under the `--gogo_out` base directory (note the missing `vendor/` folder).

```sh
~/Github/kubernetes/kubernetes ❯ awk "NR==29" vendor/k8s.io/api/flowcontrol/v1beta1/generated.proto                                                                                                         29s
option go_package = "k8s.io/api/flowcontrol/v1beta1";
~/Github/kubernetes/kubernetes ❯ protoc -I . -I vendor -I third_party/protobuf --gogo_out=tmpout vendor/k8s.io/api/flowcontrol/v1beta1/generated.proto

vendor/k8s.io/api/flowcontrol/v1beta1/generated.proto:26:1: warning: Import k8s.io/apimachinery/pkg/runtime/schema/generated.proto is unused.
vendor/k8s.io/api/flowcontrol/v1beta1/generated.proto:25:1: warning: Import k8s.io/apimachinery/pkg/runtime/generated.proto is unused.
~/Github/kubernetes/kubernetes ❯ tree tmpout
tmpout
└── k8s.io
    └── api
        └── flowcontrol
            └── v1beta1
                └── generated.pb.go

4 directories, 1 file
```

So before https://github.com/kubernetes/kubernetes/pull/105027/commits/e26c02a35f25c5ee75488a84c62a965edd8cf313 the wrong file got picked up for rewriting and adding boilerplate since `outputPath` was not referring to the file that actually got generated by `protoc-gen-gogo`. I have changed the `outputPath` such that is matches the file that gets generated by `protoc-gen-gogo`. 

See also https://github.com/gogo/protobuf/blob/226206f39bd7276e88ec684ea0028c18ec2c91ae/protoc-gen-gogo/generator/generator.go#L324-L368

liggitt commented on staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/cmd.go:
```diff
@@ -266,11 +266,17 @@ func Run(g *Generator) {
+		// output directory. Note that it is not dependend on the input path
+		// itself. All protobufPackages have their import path as go_package
+		// option set.
+		outputPath := filepath.Join(g.OutputBase, p.OutputPath())
```
I'm confused why it is correct for this to ignore `Vendored` and `VendorOutputBase`, while the construction of `path` above still honors it

fischor replied:
`path` is the input path of the current `generated.proto` file to be processed in that loop. It has been generated using the `ExecutePackage` calls in L236 and L239, either under `g.OutputBase` or `g.VendorOutputBase` depending on whether its a vendored package or not. So to calculate the `path` of that `generated.proto`, we have to differentiate if its a vendored package (`p.Vendored == true`) or not.

That `path` is passed to protoc command called in line 282: `protoc --gogo_out=<g.OutputBase> <path>`. That call will generate a new `generate.pb.go` file. Its path is `outputPath`. Previously `outputPath` was dependent on the location of the input `generated.proto` file. Now since all `generated.proto` files have a correct `go_package` option set, the output location is not dependent on the input location anymore. Instead, the output will be generated under `<g.OutputBase>/<go-package-option>/generated.pb.go`, e.g. `<g.OutputBase>/k8s.io/api/admission/v1beta1/generated.pb.go`. 

Whereas previously it was dependent on the input file location: if the `go_package` option is not a full path (i.e. does not contain a `/`) protoc-gen-gogo will generated an output file under `<g.OutputBase>/<input-file-directory>/generated.pb.go`. E.g.: for `protoc --gogo_out=<g.OutputBase> path/to/vendor/k8s.io/api/admission/v1beta1/generated.proto` a file `<g.OutputBase>/path/to/vendor/k8s.io/api/admission/v1beta1/generated.proto`, will be generated if no path as `go_package` is set.

Have you read my comment above? https://github.com/kubernetes/kubernetes/pull/105027#issuecomment-922445525 That shows an example of this behaviour.

liggitt replied:
But `<g.OutputBase>/k8s.io/api/admission/v1beta1/generated.pb.go` is not correct, right? we want `<g.VendorOutputBase>/k8s.io/api/admission/v1beta1/generated.pb.go`

liggitt commented:
https://github.com/kubernetes/kubernetes/pull/106630 was also opened which looks like it's trying to accomplish the same thing, though in a slightly different way

dims commented:
Is this PR still needed, please rebase if so (or we can close it?)

ritazh commented:
/remove-sig auth

Feel free to add us back if needed.

liggitt commented:
resolved by https://github.com/kubernetes/kubernetes/pull/106630

/close
